src/dbnode/storage/series/buffer.go

```diff
@@ -500,18 +475,205 @@ func (b *dbBuffer) FetchBlocksMetadata(
+		return
+	}
+
+	b.bucketsPool.Put(buckets)
+	delete(b.bucketsMap, xtime.ToUnixNano(blockStart))
+}
+
+type dbBufferBuckets struct {
+	buckets           []*dbBufferBucket
 	start             time.Time
-	encoders          []inOrderEncoder
-	bootstrapped      []block.DatabaseBlock
+	opts              Options
 	lastReadUnixNanos int64
-	drained           bool
+	bucketPool        *dbBufferBucketPool
+}
+
+func (b *dbBufferBuckets) resetTo(
+	start time.Time,
+	opts Options,
+	bucketPool *dbBufferBucketPool,
+) {
+	// nil all elements so that they get GC'd
+	for i := range b.buckets {
+		b.buckets[i] = nil
+	}
+	b.buckets = b.buckets[:0]
+	b.start = start
+	b.opts = opts
+	atomic.StoreInt64(&b.lastReadUnixNanos, 0)
+	b.bucketPool = bucketPool
+}
+
+func (b *dbBufferBuckets) streams(ctx context.Context, wType WriteType) []xio.BlockReader {
+	var res []xio.BlockReader
+	for _, bucket := range b.buckets {
+		if bucket.wType == wType {
+			res = append(res, bucket.streams(ctx)...)
+		}
+	}
+
+	return res
+}
+
+func (b *dbBufferBuckets) streamsLen() int {
+	res := 0
+	for _, bucket := range b.buckets {
+		res += bucket.streamsLen()
+	}
+	return res
+}
+
+func (b *dbBufferBuckets) write(
+	timestamp time.Time,
+	value float64,
+	unit xtime.Unit,
+	annotation []byte,
+	wType WriteType,
+) error {
+	return b.writableBucketCreate(wType).write(timestamp, value, unit, annotation)
+}
+
+func (b *dbBufferBuckets) merge(wType WriteType) (mergeResult, error) {
+	var res mergeResult
+	for _, bucket := range b.buckets {
+		// Only makes sense to merge buckets that are writable
+		if bucket.version == writableBucketVer && wType == bucket.wType {
+			mergeRes, err := bucket.merge()
+			if err != nil {
+				return res, nil
+			}
+			res.merges += mergeRes.merges
+		}
+	}
+
+	return res, nil
+}
+
+func (b *dbBufferBuckets) setLastRead(value time.Time) {
+	atomic.StoreInt64(&b.lastReadUnixNanos, value.UnixNano())
+}
+
+func (b *dbBufferBuckets) lastRead() time.Time {
+	return time.Unix(0, atomic.LoadInt64(&b.lastReadUnixNanos))
+}
+
+func (b *dbBufferBuckets) bootstrap(bl block.DatabaseBlock) {
+	// TODO(juchan): what is a "cold" bootstrap?
+	b.writableBucketCreate(WarmWrite).addBlock(bl)
+}
+
+func (b *dbBufferBuckets) writableBucket(wType WriteType) (*dbBufferBucket, bool) {
+	for _, bucket := range b.buckets {
+		if bucket.version == writableBucketVer && bucket.wType == wType {
+			return bucket, true
+		}
+	}
+
+	return nil, false
+}
+
+func (b *dbBufferBuckets) writableBucketCreate(wType WriteType) *dbBufferBucket {
+	bucket, exists := b.writableBucket(wType)
+
+	if exists {
+		return bucket
+	}
+
+	return b.newBucketAt(b.start, wType)
+}
+
+func (b *dbBufferBuckets) newBucketAt(
+	t time.Time,
+	wType WriteType,
+) *dbBufferBucket {
+	bucket := b.bucketPool.Get()
+	bucket.resetTo(t, wType, b.opts)
+	b.buckets = append([]*dbBufferBucket{bucket}, b.buckets...)
+	return bucket
+}
+
+func (b *dbBufferBuckets) toBlock(wType WriteType) (block.DatabaseBlock, error) {
```

Comment:
I feel like the name of this method should clarify that it only creates a block for the currently writable bucket (it ignores unwritable buckets as well as bootstrapped blocks)

Reply:
Good point, I'll rename the function to specify writable buckets only. This function actually does take into account bootstrapped blocks because it does a merge to create a block containing all data within the bucket.